docs/develop/contracts/debug-locally.md

```diff
@@ -0,0 +1,61 @@
+
+- contract execution logic does not happen in expected way;
+- storage usage is higher than expected;
+- gas cost is higher than expected.
+
+In these cases, you may want to capture the contract state and function call arguments to debug in the local sandbox. You'll make incremental changes, test locally until everything works as expected. You will also likely want to keep the test fixtures to avoid regression in future.
+
+## Obtain States and Other Parameters from Testnet/Mainnet
+
+The following things determined the contract call results:
+
+- Contract code and state;
+- Node's genesis config;
+- Context data when the transaction happens, like current epoch, block height, gas price etc;
+- Caller (signer) of the transaction and arguments.
+
+We'll need to obtain all this information to reproduce the contract call locally. This information can be divide into three categories, each of them is configured in a different way:
+
+- Genesis config
+- State records
+- Parameter provided when sending transaction
+
+### Genesis config
+
+The context data when transaction happens, like current epoch, block height, gas price etc. and testnet/mainnet specific configuration such as block time, gas amount for different kind of actions all come into this category. This parameters can only be set by start a local sandbox node from a modified version of genesis file.
+First thing you need is fetch testnet/mainnet genesis config. This can be obtained by [Genesis Config RPC](https://docs.near.org/docs/api/rpc#genesis-config). Use response file to replace sandbox's genesis.json. Now we have the same configuration as the testnet/mainnet in sandbox. But there's still something different: protocol version, block height and block timestamp. This is because genesis indicates the start of the most recent hard fork of testnet/mainnet, but many blocks and time has passed since the genesis. There's also protocol version upgrades since genesis. If your contract logic doesn't depend on these parameters replay contract method on such parameters, sandbox node should give you accurate logic, gas and storage simulation. However if you do require info like certain epoch and block height, you can modify `"genesis_height"`, `"protocol_version"` on genesis.json.
+
+Change `"genesis_time"` won't have effect. There's going to be a time travel RPC for you to tweak block timestamps. Also note a transaction cannot be controlled to be precisely scheduled of a given height and you should not have absolute block number in test logics. This is because it's not controllable in a real node so your contract logic shouldn't rely on that. More common you want to depend on how many blocks or times have passed, that's a relative number. And you may want to base on the parameters at the time block produced, such as validator set, protocol feature availability etc., these are tied to epoch rather than block, so you don't have to strictly depend on a given block number.
```

Comment:
```suggestion
Simply changing `"genesis_time"` won't have an effect. In addition, there's going to be a time travel RPC for you to tweak block timestamps. Also, note a transaction cannot be controlled to be precisely scheduled of a given height and you should not have absolute block number in test logics. This is because it's not controllable in a real node so your contract logic shouldn't rely on that. More commonly, you most likely want to depend on how many blocks or how much time has passed. This is a relative number and you may want to base on the parameters at the time the block was produced, such as a validator set, protocol feature availability, etc. These are tied to [epoch](/docs/concepts/epoch) rather than the individual block so you don't have to strictly depend on a given block number.
```